### PR TITLE
Watch online statistics changes after every play & display them in toolbar

### LINKED ISSUE #26743: Online results changes should be watched (and shown) even when not at the results screen
Currently, we have a flow for watching for statistics changes of the logged in user, but it's only acted on if the user remains at the results screen. This isn't always the case, as users regularly leave the results screen before processing completes. In such a case, two things go wrong:

- They won't be aware of the change in statistics (in stable this is [shown in a very jank way](https://github.com/ppy/osu/discussions/26627), with numbers appearing in the middle of nowhere, but it is shown and people are used to seeing it)
- Because the stats aren't updated, when they play the next beatmap it will include the difference in stats from the previous play, resulting in incorrectly high numbers. This was reported in private testing as "a loved map gave me pp" or similar.

### PR DESCRIPTION
**Pull request by bdach**
- Closes https://github.com/ppy/osu/issues/26743

https://github.com/ppy/osu/assets/20418176/9e90d971-0f4d-4ca6-8d89-31fdcfd51b7c

Code 100% conflicts with https://github.com/ppy/osu/pull/27128. And functionality possibly conflicts with https://github.com/ppy/osu/pull/27147. Dunno. I publically said I was going to do anything with this before all that happened.

### CODE REVIEW

File: osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+                titleText.Alpha = 1;
+                deltaValue.Alpha = 0;
+
+                using (BeginDelayedSequence(1000))
+                {
+                    titleText.FadeOut(250, Easing.OutQuint);
+                    deltaValue.FadeIn(250, Easing.OutQuint)
+                              .Then().Delay(1000)
+                              .Then().OnComplete(_ =>
```
Comment (Joehuu) [inline osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs:191]: The delay here seems to be broken. The delta value just rolls down immediately and it's hard to know the initial number.

Comment (bdach) [reply osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs]: fixed in https://github.com/ppy/osu/pull/27156/commits/6baa0999065adc10f3b61ee527fecc614f6fa1f1, i think

File: osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs [hunk 2]
```diff
@@ -183,11 +183,11 @@ public void Display(T before, T delta, T after)
                 titleText.Alpha = 1;
                 deltaValue.Alpha = 0;
 
-                using (BeginDelayedSequence(1000))
+                using (BeginDelayedSequence(1500))
```
Comment (peppy) [inline osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs:186]: Feels a touch sluggish. 1200 maybe.

**Comment by peppy** [comment]
Here's some changes which I think make things feel better:

```diff
diff --git a/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs b/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
index 9960dd4411..52923ea178 100644
--- a/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
+++ b/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
@@ -127,7 +127,7 @@ private void load()
                             ValuePrefix = mainValuePrefix,
                             Anchor = Anchor.CentreRight,
                             Origin = Anchor.CentreRight,
-                            Font = OsuFont.GetFont(),
+                            Font = OsuFont.GetFont().With(fixedWidth: true),
                         },
                         new Container
                         {
@@ -140,7 +140,7 @@ private void load()
                                 {
                                     Anchor = Anchor.CentreRight,
                                     Origin = Anchor.CentreRight,
-                                    Font = OsuFont.GetFont(size: 12, fixedWidth: false, weight: FontWeight.SemiBold),
+                                    Font = OsuFont.GetFont(size: 12, fixedWidth: true, weight: FontWeight.SemiBold),
                                     AlwaysPresent = true,
                                 },
                                 titleText = new OsuSpriteText
@@ -183,11 +183,11 @@ public void Display(T before, T delta, T after)
                 titleText.Alpha = 1;
                 deltaValue.Alpha = 0;
 
-                using (BeginDelayedSequence(1500))
+                using (BeginDelayedSequence(1200))
                 {
-                    titleText.FadeOut(250, Easing.OutQuint);
-                    deltaValue.FadeIn(250, Easing.OutQuint)
-                              .Then().Delay(1500)
+                    titleText.FadeOut(250, Easing.OutQuad);
+                    deltaValue.FadeIn(250, Easing.OutQuad)
+                              .Then().Delay(500)
                               .Then().Schedule(() =>
                               {
                                   mainValue.Current.Value = after;
@@ -200,9 +200,9 @@ public void Display(T before, T delta, T after)
         private partial class Counter<T> : RollingCounter<T>
             where T : struct, IEquatable<T>, IFormattable
         {
-            public const double ROLLING_DURATION = 500;
+            public const double ROLLING_DURATION = 1500;
 
-            public FontUsage Font { get; init; } = OsuFont.Default;
+            public FontUsage Font { get; init; } = OsuFont.Default.With(fixedWidth: true);
 
             public string ValuePrefix
             {
@@ -217,7 +217,13 @@ public string ValuePrefix
             private string valuePrefix = string.Empty;
 
             protected override LocalisableString FormatCount(T count) => LocalisableString.Format(@"{0}{1:N0}", ValuePrefix, count);
-            protected override OsuSpriteText CreateSpriteText() => base.CreateSpriteText().With(t => t.Font = Font);
+
+            protected override OsuSpriteText CreateSpriteText() => base.CreateSpriteText().With(t =>
+            {
+                t.Font = Font;
+                t.Spacing = new Vector2(-1.5f, 0);
+            });
+
             protected override double RollingDuration => ROLLING_DURATION;
         }
     }

```

In addition, the `Schedule` used to delay the animation seems to break a bit if clicking the test steps multiple times. I'm not sure this will happen in practice, but maybe we want to cancel the last animation when a new value change event comes in to avoid this.

**Review by peppy — COMMENTED** [review]
I post patch.

**Comment by bdach** [comment]
> the `Schedule` used to delay the animation seems to break a bit if clicking the test steps multiple times

Aware of this, which is why I attempted to initally try `OnComplete()` which [was broken](https://github.com/ppy/osu/pull/27156#discussion_r1488489432). So I don't know how one would even do this correctly at this rate.

**Comment by peppy** [comment]
Something like this probably:

```diff
diff --git a/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs b/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
index 4d3c12c8be..91fe27050e 100644
--- a/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
+++ b/osu.Game/Overlays/Toolbar/TransientUserStatisticsUpdateDisplay.cs
@@ -9,6 +9,7 @@
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Sprites;
 using osu.Framework.Localisation;
+using osu.Framework.Threading;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Sprites;
 using osu.Game.Graphics.UserInterface;
@@ -99,6 +100,8 @@ private partial class Statistic<T> : CompositeDrawable
             private Counter<T> deltaValue = null!;
             private OsuSpriteText titleText = null!;
 
+            private ScheduledDelegate? valueUpdateSchedule;
+
             [Resolved]
             private OsuColour colours { get; set; } = null!;
 
@@ -187,14 +190,17 @@ public void Display(T before, T delta, T after)
 
                 using (BeginDelayedSequence(1200))
                 {
-                    deltaValue.FadeIn(250, Easing.OutQuint)
-                              .Then().Delay(1500)
-                              .Then().Schedule(() =>
-                              {
-                                  mainValue.Current.Value = after;
-                                  deltaValue.Current.SetDefault();
-                              });
                     titleText.FadeOut(250, Easing.OutQuad);
+                    deltaValue.FadeIn(250, Easing.OutQuad);
+
+                    using (BeginDelayedSequence(250))
+                    {
+                        valueUpdateSchedule = Schedule(() =>
+                        {
+                            mainValue.Current.Value = after;
+                            deltaValue.Current.SetDefault();
+                        });
+                    }
                 }
             }
         }

```

**Comment by bdach** [comment]
Have applied, thanks.

**Review by peppy — APPROVED** [review]

**Pull request merged**